File: src/backend/collide.rs
// Names a paste or a drop would land on: the question asked first, and the one choice the transfer carries.
use crate::backend::icons::Names;
use crate::backend::menu_actions::Selected;
use crate::backend::mime::Db;
use crate::backend::ops::free_copy_path;
use crate::backend::opsdispatch::Ops;
use crate::backend::opsreq::{usable_dest, OpMsg, ALREADY_THERE};
use crate::backend::trash;
use crate::backend::undo::{ItemIdentity, Step};
use crate::error::FleaError;
use crate::json::{escape, field_str, field_usize};
use std::cell::OnceCell;
use std::collections::{HashMap, HashSet};
use std::os::unix::fs::MetadataExt;
use std::path::{Path, PathBuf};
use std::sync::Arc;

// The card lists this many names and says how many more there are, so the answer never carries more.
pub const SHOWN: usize = 3;
// The word Duplicate already uses, so a kept copy and a duplicate are named by one rule.
const KEEP_WORD: &str = "copy";
pub const TRASH_REFUSED: &str = "the item already there could not be moved to Trash, so nothing was replaced";
pub const HOLDS_SOURCE: &str = "the item already there holds the one being moved in, so it was not replaced";
pub const LINKS_HERE: &str = "the incoming link points at the item already there, so it was not replaced";
pub const NO_FREE_NAME: &str = "every copy name for this item is already taken";
// copyfile's own word for a cancelled item, which is what the transfer counts a cancel by.
const CANCELLED: &str = "cancelled";

// The operator's one answer, for every name the question listed.
#[derive(Clone, Copy, Debug, PartialEq)]
pub enum Collide {
    Refuse,
    Keep,
    Replace,
    Skip,
}

impl Collide {
    // Anything that is not one of the three choices refuses, so a malformed word can never replace.
    fn from_word(word: &str) -> Collide {
        match word {
            "keep" => Collide::Keep,
            "replace" => Collide::Replace,
            "skip" => Collide::Skip,
            _ => Collide::Refuse,
        }
    }
}

// What the question saw: each colliding source, and the item that held its name in dest at that moment.
pub struct Question {
    id: usize,
    dest: PathBuf,
    seen: HashMap<PathBuf, ItemIdentity>,
    menu: Option<MenuCapture>,
}

// A menu's selection as its question saw it: Copy to closes its dialog, which expires the live one, before the answer lands.
struct MenuCapture {
    id: usize,
    items: Vec<Selected>,
    destination: Option<Selected>,
}

// The transfer request's two fields, before the question they name has been looked up.
#[derive(Clone, Debug, Default, PartialEq)]
pub struct Ask {
    word: Option<String>,
    id: usize,
}

impl Ask {
    // Sample input: {"c":"transfer","op":"copy","paths":["/home/gm/a.png"],"dest":"/home/gm/Pictures","collide":"replace","collideId":7}
    pub fn parse(line: &str) -> Ask {
        Ask { word: field_str(line, "collide"), id: field_usize(line, "collideId").unwrap_or(0) }
    }

    // The choice covers only the question this request names, asked about this same destination.
    pub fn policy(self, question: Option<Question>, dest: &Path) -> Policy {
        let seen = match question {
            Some(q) if self.answers(&q, dest) => q.seen,
            _ => HashMap::new(),
        };
        Policy { choice: self.word.as_deref().map(Collide::from_word), seen, ..Policy::default() }
    }

    fn answers(&self, question: &Question, dest: &Path) -> bool {
        self.id != 0 && question.id == self.id && question.dest == dest
    }
}

// The selection a menu transfer runs on: the one its question captured, else the menu's live one.
pub(crate) fn menu_sources(ops: &Ops, id: usize, dest: &str, collide: &Ask) -> Result<(Vec<Selected>, Option<Selected>), String> {
    let captured = ops.question.as_ref().filter(|q| collide.answers(q, Path::new(dest))).and_then(|q| q.menu.as_ref());
    if let Some(menu) = captured.filter(|menu| menu.id == id) {
        return Ok((menu.items.clone(), menu.destination.clone()));
    }
    ops.menuactions.as_ref().ok_or_else(|| "Menu selection expired; reopen the menu.".to_string())
        .and_then(|menu| Ok((menu.selection(id)?, menu.provider_destination(id, Path::new(dest))?)))
}

// An expired selection captures nothing, so the question answers no collision and the transfer its own error.
fn capture(ops: &Ops, id: usize, dest: &str) -> Option<MenuCapture> {
    let menu = ops.menuactions.as_ref()?;
    Some(MenuCapture { id, items: menu.selection(id).ok()?, destination: menu.provider_destination(id, Path::new(dest)).ok()? })
}

// No choice at all is today's transfer exactly; any choice also turns on the same-folder rule.
#[derive(Default)]
pub struct Policy {
    choice: Option<Collide>,
    seen: HashMap<PathBuf, ItemIdentity>,
    batch: Vec<PathBuf>,
    // Every source of the batch and every folder above it, resolved on the first Replace, so one lookup says whether a trash would take a source.
    held: OnceCell<HashSet<PathBuf>>,
}

// Where one item goes; replace says what holds that path now goes to Trash first.
#[derive(Debug, PartialEq)]
pub enum Place {
    Land { to: PathBuf, replace: bool },
    Skip,
    Refuse(String),
}

impl Policy {
    // Replace checks each item against the whole batch, so the batch is kept only when there is a Replace to check.
    pub fn for_batch(mut self, paths: &[String]) -> Policy {
        if self.choice == Some(Collide::Replace) {
            self.batch = paths.iter().map(PathBuf::from).collect();
        }
        self
    }

    // A skipped item copies nothing, so the sweep leaves its bytes out of the batch's total.
    pub fn skips(&self, src: &Path) -> bool {
        self.choice == Some(Collide::Skip) && self.seen.contains_key(src)
    }

    // here says the item already lives in dest; a name the question did not see is left to the exclusive create, which refuses it.
    pub fn place(&self, src: &Path, dst: PathBuf, here: bool, moving: bool) -> Place {
        let Some(choice) = self.choice else {
            return if here { Place::Refuse(ALREADY_THERE.to_string()) } else { Place::Land { to: dst, replace: false } };
        };
        if here {
            // A move onto itself changes nothing, and a copy into its own folder is a Duplicate.
            return if moving { Place::Skip } else { keep_both(dst) };
        }
        let Ok(current) = ItemIdentity::inspect(&dst) else { return Place::Land { to: dst, replace: false } };
        if !self.seen.get(src).is_some_and(|seen| seen.same_item(&current)) {
            return Place::Land { to: dst, replace: false };
        }
        match choice {
            Collide::Refuse => Place::Land { to: dst, replace: false },
            Collide::Keep => keep_both(dst),
            Collide::Skip => Place::Skip,
            Collide::Replace => match self.guard(&dst, src) {
                Some(reason) => Place::Refuse(reason.to_string()),
                None => Place::Land { to: dst, replace: true },
            },
        }
    }

    // What a trash of dst would take with it: any source this batch names, or the item an incoming link resolves to.
    fn guard(&self, dst: &Path, src: &Path) -> Option<&'static str> {
        let there = resolved_parent(dst)?;
        if self.held.get_or_init(|| held_sources(&self.batch)).contains(&there) {
            return Some(HOLDS_SOURCE);
        }
        let dest = dst.parent()?;
        links_into(src, dest, &there).then_some(LINKS_HERE)
    }
}

// Each source and its ancestors; an ancestor already in the set had its own ancestors added with it.
fn held_sources(batch: &[PathBuf]) -> HashSet<PathBuf> {
    let mut held = HashSet::new();
    // A selection shares a folder or a few, so each folder is resolved once rather than once per item.
    let mut folders: HashMap<&Path, Option<PathBuf>> = HashMap::new();
    for src in batch {
        let (Some(parent), Some(name)) = (src.parent(), src.file_name()) else { continue };
        let folder = folders.entry(parent).or_insert_with(|| parent.canonicalize().ok());
        let Some(resolved) = folder.as_ref().map(|folder| folder.join(name)) else { continue };
        for ancestor in resolved.ancestors() {
            if !held.insert(ancestor.to_path_buf()) {
                break;
            }
        }
    }
    held
}

// A link lands under its own text in dest, so it resolves into there either now or once it sits at dest.
fn links_into(src: &Path, dest: &Path, there: &Path) -> bool {
    let Ok(text) = std::fs::read_link(src) else { return false };
    let now = src.canonicalize().is_ok_and(|target| target.starts_with(there));
    now || resolved_parent(&dest.join(text)).is_some_and(|landed| landed.starts_with(there))
}

fn keep_both(dst: PathBuf) -> Place {
    match free_copy_path(&dst, KEEP_WORD) {
        Some(free) => Place::Land { to: free, replace: false },
        None => Place::Refuse(NO_FREE_NAME.to_string()),
    }
}

// The parent resolved and the last component left as it is, so a symlink names itself and not its target.
fn resolved_parent(path: &Path) -> Option<PathBuf> {
    Some(path.parent()?.canonicalize().ok()?.join(path.file_name()?))
}

// The item already lives in dest: its name there is the item itself, through a symlinked folder too.
pub fn already_there(src: &Path, name: &str, dst: &Path, dest_real: &Path) -> bool {
    let src_here = match src.parent() {
        Some(parent) => parent.canonicalize().unwrap_or_else(|_| parent.to_path_buf()).join(name),
        None => src.to_path_buf(),
    };
    dst == src || dest_real.join(name) == src_here
}

// Replace is Trash and then the transfer, journaled as one entry: undo removes the new item, then restores the old.
pub fn replacing(dst: &Path, steps: &mut Vec<Step>, land: impl FnOnce(&mut Vec<Step>) -> Result<(), FleaError>) -> Result<(), FleaError> {
    let (mut entries, failed) = trash::trash(&[dst.to_path_buf()]);
    if failed != 0 || entries.len() != 1 {
        return Err(FleaError { where_: "transfer".into(), path: dst.to_string_lossy().into(), msg: TRASH_REFUSED.into() });
    }
    let entry = entries.remove(0);
    let at = steps.len();
    steps.push(Step::Trashed(entry.clone()));
    let mut outcome = land(steps);
    if let Err(error) = &mut outcome {
        put_back(&entry, dst, steps, at, error);
    }
    outcome
}

// Nothing took the name, a cancel included, so the old item goes straight back instead of waiting on an undo.
fn put_back(entry: &trash::Entry, dst: &Path, steps: &mut Vec<Step>, at: usize, error: &mut FleaError) {
    let vacant = matches!(dst.symlink_metadata(), Err(e) if e.kind() == std::io::ErrorKind::NotFound);
    if !vacant || steps.len() != at + 1 {
        return;
    }
    match trash::restore(entry) {
        Ok(()) => steps.truncate(at),
        // A cancel keeps its own word, which is what counts it; the step stays, so undo can still restore it.
        Err(_) if error.msg == CANCELLED => {}
        Err(restore) => error.msg.push_str(&format!("; the item it replaced is still in Trash ({})", restore.msg)),
    }
}

// One colliding source as the card draws it: its name, and the kind its mark is chosen from.
struct Shown {
    name: String,
    dir: bool,
    mode: u32,
}

// The question itself, read-only: which sources would land on a name dest already holds.
fn ask(id: usize, paths: &[String], dest: &str) -> (Question, Vec<Shown>) {
    let mut question = Question { id, dest: PathBuf::from(dest), seen: HashMap::new(), menu: None };
    let mut shown = Vec::new();
    // An unusable destination asks nothing: the transfer that follows answers its own error.
    let Ok(dest_path) = usable_dest(dest) else { return (question, shown) };
    let dest_real = dest_path.canonicalize().unwrap_or_else(|_| dest_path.clone());
    for raw in paths {
        let src = Path::new(raw);
        if !src.is_absolute() {
            continue;
        }
        let Some(name) = src.file_name().and_then(|n| n.to_str()) else { continue };
        let Ok(meta) = src.symlink_metadata() else { continue };
        let dst = dest_path.join(name);
        // Same-folder items never ask: a copy keeps both and a move stays put, see Policy::place.
        if already_there(src, name, &dst, &dest_real) {
            continue;
        }
        let Ok(there) = ItemIdentity::inspect(&dst) else { continue };
        if question.seen.insert(src.to_path_buf(), there).is_none() && shown.len() < SHOWN {
            shown.push(Shown { name: name.to_string(), dir: meta.is_dir(), mode: meta.mode() });
        }
    }
    (question, shown)
}

// Sample output: {"t":"collisions","id":7,"total":1,"names":[{"n":"a.png","d":false,"i":"image-x-generic"}]}
fn collisions_line(id: usize, total: usize, shown: &[Shown], mime: &Db, icons: &Names) -> String {
    let names: Vec<String> = shown.iter().map(|s| format!(r#"{{"n":"{}","d":{},"i":"{}"}}"#,
        escape(&s.name), s.dir, escape(icons.icon_for(mime.lookup(&s.name), s.dir, s.mode)))).collect();
    format!(r#"{{"t":"collisions","id":{},"total":{},"names":[{}]}}"#, id, total, names.join(","))
}

// Measured 0.9 s for 100,000 local sources, a stall per network round trip on a mount, so the question runs beside the loop.
pub(crate) fn ask_beside(ops: &mut Ops, id: usize, menu_id: usize, named: Vec<String>, dest: &str, mime: &Arc<Db>, icons: &Arc<Names>) {
    // Captured here and not on the thread, because the close that expires the live selection may be the very next request.
    let menu = if menu_id == 0 { None } else { capture(ops, menu_id, dest) };
    let paths = match (&menu, menu_id) {
        (Some(menu), _) => menu.items.iter().map(|item| item.path.to_string_lossy().to_string()).collect(),
        (None, 0) => named,
        (None, _) => Vec::new(),
    };
    ops.asked += 1;
    let (turn, tx, dest, mime, icons) = (ops.asked, ops.tx.clone(), dest.to_string(), Arc::clone(mime), Arc::clone(icons));
    std::thread::spawn(move || {
        let (mut question, shown) = ask(id, &paths, &dest);
        question.menu = menu;
        let line = collisions_line(id, question.seen.len(), &shown, &mime, &icons);
        let _ = tx.send(OpMsg::Asked { turn, question, line });
    });
}

// The latest question asked is the one kept, so an earlier one that finishes later cannot stand in for it; the one transfer that names it spends it.
pub(crate) fn landed(ops: &mut Ops, turn: usize, question: Question) {
    if turn == ops.asked {
        ops.question = Some(question);
    }
}

#[cfg(test)]
#[path = "collide_tests.rs"]
mod tests;
#[cfg(test)]
#[path = "collide_replace_tests.rs"]
mod replace_tests;

File: tests/js/collide.js
.import "../../ui/js/Anchor.js" as Anchor
.import "../../ui/js/Collide.js" as Collide
.import "../../ui/js/Drag.js" as Drag
.import "../../ui/js/Ops.js" as Ops
.import "../../ui/js/Swap.js" as Swap
.import "collidefixture.js" as Fixture

// The collision card's decisions, and that every paste and drop asks it before anything is sent.

function names(list) {
    return list.map(function (n) { return { n: n, d: false, i: "image-x-generic" } })
}

function run(check) {
    // The board's two titles, word for word.
    check("one name says which and where", Collide.title(names(["screenshot.png"]), 1, "/home/gm/Pictures"),
          "screenshot.png already exists in Pictures")
    check("several say how many", Collide.title(names(["a", "b", "c"]), 3, "/home/gm/Downloads"),
          "3 items already exist in Downloads")
    check("a count past the list still says the whole count", Collide.title(names(["a", "b", "c"]), 12, "/home/gm/Downloads"),
          "12 items already exist in Downloads")
    check("the root has no leaf, so it names itself", Collide.folderName("/"), "/")
    check("a trailing name is the folder", Collide.folderName("/home/gm/Pictures"), "Pictures")

    // The row under the list: absent while every collision is on it.
    check("three of three needs no more line", Collide.more(3, 3), "")
    check("one past the list", Collide.more(4, 3), "and 1 more")
    check("many past it", Collide.more(40, 3), "and 37 more")
    check("the explanation is one sentence", Collide.EXPLAIN, "Replaced items go to Trash, and Undo restores them.")

    // Focus: Keep both first, h and l stop at the ends, Tab and Backtab come round.
    check("the card opens on Keep both", Collide.START, "keep")
    check("the buttons read left to right", Collide.BUTTONS.map(function (b) { return Collide.LABELS[b] }).join("|"),
          "Cancel|Skip|Keep both|Replace")
    check("l steps right", Collide.moved("keep", Qt.Key_L), "replace")
    check("Right is l", Collide.moved("skip", Qt.Key_Right), "keep")
    check("l stops at Replace", Collide.moved("replace", Qt.Key_L), "replace")
    check("h steps left", Collide.moved("keep", Qt.Key_H), "skip")
    check("Left is h", Collide.moved("skip", Qt.Key_Left), "cancel")
    check("h stops at Cancel", Collide.moved("cancel", Qt.Key_H), "cancel")
    check("Tab comes round from Replace", Collide.moved("replace", Qt.Key_Tab), "cancel")
    check("Backtab comes round from Cancel", Collide.moved("cancel", Qt.Key_Backtab), "replace")
    check("Tab walks every button left to right", ["cancel", "skip", "keep"].map(function (b) { return Collide.moved(b, Qt.Key_Tab) }).join("|"),
          "skip|keep|replace")
    check("and Backtab walks them back", ["skip", "keep", "replace"].map(function (b) { return Collide.moved(b, Qt.Key_Backtab) }).join("|"),
          "cancel|skip|keep")
    check("any other key leaves the focus", Collide.moved("skip", Qt.Key_J), "skip")
    check("Enter takes the focused choice", Collide.activates(Qt.Key_Return) && Collide.activates(Qt.Key_Enter), true)
    check("and so does Space", Collide.activates(Qt.Key_Space), true)
    check("Escape is not an activation, it cancels", Collide.activates(Qt.Key_Escape), false)

    // The question names the same sources the transfer will.
    var byPath = { c: "transfer", op: "copy", paths: ["/a/x.png"], dest: "/b" }
    check("a paste asks about its paths", JSON.stringify(Collide.question(byPath, null, 4)),
          JSON.stringify({ c: "collisions", id: 4, dest: "/b", paths: ["/a/x.png"] }))
    var byRows = { c: "transfer", op: "move", rows: [2, 3], dest: "/d/omarchy" }
    check("a row drop asks about its rows", JSON.stringify(Collide.question(byRows, null, 5)),
          JSON.stringify({ c: "collisions", id: 5, dest: "/d/omarchy", rows: [2, 3] }))
    var shelf = { c: "transfer", op: "", paths: [], dest: "/e", shelf: "tok" }
    check("a shelf drop asks about the paths its drag carries", JSON.stringify(Collide.question(shelf, ["/s/a.txt"], 6)),
          JSON.stringify({ c: "collisions", id: 6, dest: "/e", paths: ["/s/a.txt"] }))
    var copyTo = { c: "transfer", op: "copy", menuId: 31, dest: "/f" }
    check("Copy to asks about the menu's own selection", JSON.stringify(Collide.question(copyTo, null, 7)),
          JSON.stringify({ c: "collisions", id: 7, dest: "/f", menuId: 31 }))
    var dropbox = { c: "transfer", op: "move", rows: [1, 2], dest: "/dropbox", menuId: 32 }
    check("and so does Move to Dropbox, whose rows the backend never reads beside a menu",
          JSON.stringify(Collide.question(dropbox, null, 8)), JSON.stringify({ c: "collisions", id: 8, dest: "/dropbox", menuId: 32 }))
    check("the answer rides on the transfer it was asked for", JSON.stringify(Collide.transfer(byPath, "replace", 4)),
          JSON.stringify({ c: "transfer", op: "copy", paths: ["/a/x.png"], dest: "/b", collide: "replace", collideId: 4 }))
    check("and the waiting request itself is not changed", byPath.collide, undefined)
    check("nothing colliding still says refuse", Collide.NONE, "refuse")
    check("a question still in flight refuses a second ask out loud", Collide.refusal(false, byPath), Collide.WAITING)
    check("and so does an open card", Collide.refusal(true, byPath), Collide.WAITING)
    check("even with nothing waiting behind it", Collide.refusal(true, null), Collide.WAITING)
    check("with nothing waiting a question may go", Collide.refusal(false, null), "")

    // ui/CollideHost.qml ask() stamps rows read under listing 4; a rows line in listing 5 lands before the choice sends them.
    var captured = Collide.waiting(byRows, 4)
    check("rows the card holds keep the numbering they were read in", captured.listing, 4)
    check("and the question about them names it", Collide.question(captured, null, 9).listing, 4)
    check("so ui/Backend.qml send() in listing 5 still sends the transfer as 4, which the backend refuses",
          Swap.named(Collide.transfer(captured, "replace", 9), 5).listing, 4)
    check("an unstamped row request takes the numbering in force", Swap.named(byRows, 5).listing, 5)
    check("while a request naming no rows, or one before any rows line, is never stamped",
          Swap.named(byPath, 4).listing + "|" + Swap.named(byRows, 0).listing, "undefined|undefined")
    check("and the request the card was handed is not changed", byRows.listing, undefined)
    check("a menu's transfer is the menu's own capture, so its unread rows take the numbering in force at the send",
          Collide.waiting(dropbox, 4).listing + "|" + Swap.named(Collide.transfer(Collide.waiting(dropbox, 4), "keep", 8), 5).listing,
          "undefined|5")

    // A paste asks rather than sends, and leaves a cut on the clipboard until the transfer goes out.
    var asked = []
    var pane = { path: "/dest", clipboard: { paths: ["/src/a.txt"], moving: true },
                 collide: { ask: function (request, probe, cut) { asked.push([request, probe, cut]) } } }
    Ops.paste(pane)
    check("a paste asks the question first", asked.length === 1 ? asked[0][0].c + " " + asked[0][0].op + " " + asked[0][0].dest : "none",
          "transfer move /dest")
    check("a cut paste is marked as spending the cut", asked[0][2], true)
    check("and the cut stays until the answer sends it", pane.clipboard.paths.length, 1)
    pane.clipboard = { paths: ["/src/a.txt"], moving: false }
    Ops.paste(pane)
    check("a copy paste asks a copy that spends nothing", asked[1][0].op + " " + asked[1][2], "copy false")

    // Drops ask too, and a shelf drop asks about the paths its drag carries while its token names what moves.
    var sent = []
    var answer = false
    asked = []
    var drops = { path: "/d", rowFor: function () { return { n: "omarchy", d: true } }, join: function (a, b) { return a + "/" + b },
                  backend: { send: function (msg) { sent.push(msg) } },
                  collide: { ask: function (request, probe) { asked.push([request, probe]); return answer } } }
    check("a row drop answers the card's refusal", Drag.drop(drops, [2], 0, false), false)
    check("a path drop does too", Drag.dropInto(drops, "", ["file:///x/a.txt"], "/e", 0), false)
    check("a shelf drop asks with the paths its drag carries",
          Drag.dropInto(drops, "", ["file:///s/a.txt"], "/e", 0, "tok\nmove") + " " + (asked[2] ? JSON.stringify(asked[2][1]) : "not asked"), "false [\"/s/a.txt\"]")
    answer = true
    check("and each answers the card's acceptance, not a false of its own",
          Drag.drop(drops, [2], 0, false, 7) + " " + Drag.dropInto(drops, "", ["file:///x/a.txt"], "/e", 0), "true true")
    check("a row drop names the listing it is handed rather than none", asked[3][0].listing, 7)
    check("every drop asked and none went straight to the backend", asked.length + " " + sent.length, "5 0")
    var lifted = { c: "transfer", op: "move", rows: [2], dest: "/d/omarchy", listing: 7 }
    check("a request already stamped, a drag lifted in 7, keeps 7 when the card asks while the pane holds 8",
          Collide.waiting(lifted, 8).listing + "|" + Collide.question(Collide.waiting(lifted, 8), null, 10).listing, "7|7")

    wired(check)
}

// The real ui/FileDrag.qml, ui/RowDrag.qml and ui/CollideHost.qml: a row lifted in 7, a re-list to 8 while it is up, the drop, Replace.
function wired(check) {
    var at = Fixture.scene(7)
    var backend = at.backend, p = at.pane, session = at.session, target = at.target
    check("a pane at rest holds no watched re-read back", Anchor.busy(p), false)
    var taken = false
    // Every payload field is fixed once the lift sets its feedback, and offscreen's platform drag returns at once, so the drop lands here.
    session.feedbackChanged.connect(function () {
        if (session.feedback === null)
            return
        backend.heldListing = 8
        taken = target.dropped(session.dragMime[Drag.ROWS_MIME], [], "")
    })
    session.liftBegan(0, { modifiers: Qt.NoModifier })
    check("a row lifted in 7 and dropped on a folder after a re-list to 8 asks about its rows in 7",
          taken + " " + JSON.stringify(backend.sent),
          "true " + JSON.stringify([{ c: "collisions", id: 1, dest: "/d/omarchy", rows: [0], listing: 7 }]))
    check("the watched re-read waits while the card holds the transfer", Anchor.busy(p), true)
    var released = ""
    // The first moment it lets go, since a var property signals every null written to it.
    p.collide.pendingChanged.connect(function () {
        if (p.collide.pending === null && released === "")
            released = Fixture.verbs(backend) + "|" + Anchor.busy(p)
    })
    p.collide.decide("replace")
    var transfer = backend.sent[1] || {}
    check("Replace sends it in 7 while 8 is in force, so the backend refuses it rather than moving another file",
          transfer.c + " " + JSON.stringify(transfer.rows) + " " + transfer.listing + " " + transfer.collide, "transfer [0] 7 replace")
    check("and writes it before the card lets go of it, which is what frees the re-read", released, "collisions,transfer|false")
    p.collide.ask({ c: "transfer", op: "move", paths: ["/s/a.txt"], dest: "/d" }, null, true)
    p.collide.decide("cancel")
    check("a Cancel writes only its question, lets go, and leaves the cut on the clipboard",
          Fixture.verbs(backend) + "|" + (p.collide.pending === null) + "|" + p.clipboard.paths.length, "collisions,transfer,collisions|true|1")
    at.parent.destroy()
}
